### PR TITLE
MRG improve how page

### PR DESCRIPTION
**Pull request by agramfort**

### CODE REVIEW

File: doc/how.rst [hunk 1]
```diff
@@ -3,87 +3,160 @@
...
 1. Objective
 ------------
 
-The objective function is defined through a Python class that
-allows to evaluate the function that should be minimized by solvers.
-A objective should define a `set_data` method that allows
-to specify the data and a `compute` method that allows
-to evaluate the objective for a given value of the iterate.
+The **objective function** is defined through a Python class.
+This class allows to evaluate the objective function to be minimized
```
Comment (mathurinm) [inline doc/how.rst:40]: is it necessarily minimized (eg left out loss in stochastic bench) ? 
"This class allows to monitor the quantities of interest along iterations of the solvers" ?

File: doc/how.rst [hunk 2]
```diff
@@ -3,87 +3,160 @@
 Write a benchmark
 =================
 
-.. contents:: Contents
-   :local:
-   :depth: 2
+A benchmark consists in three elements: an objective_ function,
```
Comment (mathurinm) [inline doc/how.rst:6]: consists of ?

File: doc/how.rst [hunk 3]
```diff
@@ -3,87 +3,160 @@
...
 across problem dimensions.
 
-Example of parametrized dataset
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Example of parametrized simulated dataset
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Using simulated data:
+Sometimes one wants to test the solvers for variants of the same dataset.
+For example, one may want to change the dataset size, the noise level etc.
```
Comment (mathurinm) [inline doc/how.rst:84]: comma before etc

File: doc/how.rst [hunk 4]
```diff
@@ -3,87 +3,160 @@
...
+To be able to specify parameters to get a dataset you can use a class
+attribute called `parameters`. This parameter must be a dictionary
+whose keys can be passed to the `__init__` of the class. The benchopt
```
Comment (mathurinm) [inline doc/how.rst:87]: The > Then ?

File: doc/how.rst [hunk 5]
```diff
@@ -3,87 +3,160 @@
...
+.. _solvers:
+
 3. Solvers
 ----------
 
-Solver requires to define a `set_objective` that is constructed
-from the union of the data and the objective parameters.
+Solver requires to define three methods:
+
+   - **set_objective**: This method expects as input the union of what is returned
```
Comment (mathurinm) [inline doc/how.rst:99]: union is unclear to me here

File: doc/how.rst [hunk 6]
```diff
@@ -3,87 +3,160 @@
...
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. _python_conda_solvers:
+
+Python solver from Conda package
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Using a package available from conda:
+If you want to compare a solver available on conda you can specify
+how this solver needs to be installed and how to call it.
+The `install_cmd` class variable needs to be set to `conda`
```
Comment (mathurinm) [inline doc/how.rst:136]: `"conda"` ?

File: doc/how.rst (hunk 2 above)
Comment (tomMoral) [reply doc/how.rst:6]: `is composed of`?

File: doc/how.rst (hunk 1 above)
Comment (tomMoral) [inline doc/how.rst:40]:
```suggestion
This class allows to evaluate the objective function on the output produced by the solvers.
```

File: doc/how.rst [hunk 7]
```diff
@@ -3,87 +3,160 @@
...
-allows to evaluate the function that should be minimized by solvers.
-A objective should define a `set_data` method that allows
-to specify the data and a `compute` method that allows
-to evaluate the objective for a given value of the iterate.
+The **objective function** is defined through a Python class.
+This class allows to evaluate the objective function to be minimized
+by the solvers. An objective class should define 3 methods:
+
+  - **set_data**: it allows to specify the data. See the data as a dictionary
+    of Python variables without any constraint.
```
Comment (tomMoral) [inline doc/how.rst:44]:
```suggestion
  - **set_data**: it allows to specify the data. See the data as a dictionary
  - **set_data**: it allows to specify the data used to compute the objective.
  See the data as a dictionary of Python variables without any constraint.
```

File: doc/how.rst [hunk 8]
```diff
@@ -3,87 +3,160 @@
...
+  - **compute**: it allows to evaluate the objective for a given value
+    of the iterate. This method should take only one parameter.
```
Comment (tomMoral) [inline doc/how.rst:46]:
```suggestion
    of the iterate. This method should take only one parameter, the output
    returned by the solver. All other parameters should be stored in the
    class in `set_data`.
```

File: doc/how.rst [hunk 9]
```diff
@@ -3,87 +3,160 @@
...
-A dataset defines what can be passed to the `set_data` method
-of an objective. A dataset should implement a `get_data` method
-whose output can be passed to the `set_data` of an objective.
+A dataset defines what can be passed to an objective. More specifically,
+a dataset should implement one method:
 
-Example
-~~~~~~~
+   - **get_data**: A method whose output consists of two things. First it outputs
+     the dimension of the optization problem (size of the iterates). Second it
```
Comment (tomMoral) [inline doc/how.rst:66]:
```suggestion
     the dimension of the optimization problem (size of the iterates). Second it
```

File: doc/how.rst [hunk 10]
```diff
@@ -3,87 +3,160 @@
...
+     outputs a dictionary that can be passed as `**kwargs` to
```
Comment (tomMoral) [inline doc/how.rst:67]:
```suggestion
     outputs a dictionary that can be passed as keyword arguments `**kwargs` to
```

File: doc/how.rst [hunk 11]
```diff
@@ -3,87 +3,160 @@
...
 
-Example of parametrized dataset
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Example of parametrized simulated dataset
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Using simulated data:
+Sometimes one wants to test the solvers for variants of the same dataset.
+For example, one may want to change the dataset size, the noise level etc.
+To be able to specify parameters to get a dataset you can use a class
```
Comment (tomMoral) [inline doc/how.rst:85]:
```suggestion
To be able to specify parameters to get a dataset, you can use a class
```

File: doc/how.rst (hunk 4 above)
Comment (tomMoral) [reply doc/how.rst:87]: This is actually not necessary to be able to pass them to the class `__init__`. They will be stored anyway as `self.key = value`

File: doc/how.rst [hunk 12]
```diff
@@ -3,87 +3,160 @@
...
+
 3. Solvers
 ----------
 
-Solver requires to define a `set_objective` that is constructed
-from the union of the data and the objective parameters.
+Solver requires to define three methods:
+
+   - **set_objective**: This method expects as input the union of what is returned
+     by the **get_data** methods of both the dataset and the objective.
```
Comment (tomMoral) [inline doc/how.rst:100]:
```suggestion
   - **set_objective**: This method will be passed the dictionary returned by `to_dict` in the objective
   as keyword arguments `**kwargs`. The goal of this method is to provide all necessary information
   to the solver so it can optimize the objective function.
```

File: doc/how.rst (hunk 6 above)
Comment (tomMoral) [reply doc/how.rst:136]: Here, this also work with `pip` installable solvers.

File: doc/how.rst [hunk 13]
```diff
@@ -3,87 +3,160 @@
...
 
 .. literalinclude:: ../benchopt/tests/test_benchmarks/dummy_benchmark/solvers/sklearn.py
 
-Example of a solver available from a package available from source
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. _python_source_solvers:
 
-Using a package available from source:
+Python solver from source distrution
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (tomMoral) [inline doc/how.rst:145]:
```suggestion
Python solver from source distribution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: doc/how.rst [hunk 14]
```diff
@@ -3,87 +3,160 @@
...
+
+.. note::
+    A package available from source may require a C++
+    or Fortran compiler.
```
Comment (tomMoral) [inline doc/how.rst:149]: Je mettrai celui la en dernier, c'est le plus complexe.

**Review by tomMoral — COMMENTED** [review]
A first batch of comments.
